File: optimization-heap-operations-problems-enhanced/src/main/java/com/codinginterview/heapoperations/utils/ListNode.java
```java
package com.codinginterview.heapoperations.utils;

/**
 * Definition for singly-linked list.
 * This class is a standard utility for linked list problems.
 */
public class ListNode {
    public int val;
    public ListNode next;

    public ListNode() {}

    public ListNode(int val) {
        this.val = val;
    }

    public ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    /**
     * Helper method to convert an array to a ListNode chain.
     *
     * @param arr The array of integers.
     * @return The head of the linked list.
     */
    public static ListNode fromArray(int[] arr) {
        if (arr == null || arr.length == 0) {
            return null;
        }
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        for (int val : arr) {
            current.next = new ListNode(val);
            current = current.next;
        }
        return dummy.next;
    }

    /**
     * Helper method to convert a ListNode chain to an array.
     *
     * @param head The head of the linked list.
     * @return An array of integers representing the linked list.
     */
    public static int[] toArray(ListNode head) {
        if (head == null) {
            return new int[0];
        }
        java.util.List<Integer> list = new java.util.ArrayList<>();
        ListNode current = head;
        while (current != null) {
            list.add(current.val);
            current = current.next;
        }
        return list.stream().mapToInt(i -> i).toArray();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        ListNode current = this;
        while (current != null) {
            sb.append(current.val);
            if (current.next != null) {
                sb.append(" -> ");
            }
            current = current.next;
        }
        return sb.toString();
    }
}
```

File: data-structures-heap-operations-pro/src/data-structures/ListNode.ts
```typescript
/**
 * src/data-structures/ListNode.ts
 * 
 * Defines a simple singly-linked list node structure.
 * This is used as a helper data structure for problems involving linked lists,
 * such as "Merge K Sorted Lists".
 */
export class ListNode {
    val: number;
    next: ListNode | null;

    /**
     * Constructs a new ListNode.
     * @param val The value of the node.
     * @param next The next node in the list, or null if it's the last node.
     */
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }

    /**
     * Helper to create a linked list from an array.
     * @param arr The array of numbers to convert into a linked list.
     * @returns The head of the created linked list, or null if the array is empty.
     */
    static fromArray(arr: number[]): ListNode | null {
        if (!arr || arr.length === 0) {
            return null;
        }
        const head = new ListNode(arr[0]);
        let current = head;
        for (let i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }

    /**
     * Helper to convert a linked list to an array.
     * @param head The head of the linked list.
     * @returns An array of numbers representing the linked list.
     */
    static toArray(head: ListNode | null): number[] {
        const arr: number[] = [];
        let current = head;
        while (current) {
            arr.push(current.val);
            current = current.next;
        }
        return arr;
    }
}
```